Add class CircuitsMatroid

gmou3 opened the pull request:
<!-- ^^^^^
Please provide a concise, informative and self-explanatory title.
Don't put issue numbers in there, do this in the PR body below.
For example, instead of "Fixes #1234" use "Introduce new method to calculate 1+1"
-->
<!-- Describe your changes here in detail -->
Added a subclass `CircuitsMatroid` (of the main `Matroid` class). This enables the storage of a matroid based on its list of circuits, the verification using the circuit axioms, etc. The circuits description can be very succinct and it can facilitate efficient implementations of algorithms (one such example is the aforementioned verification).

Up till now, when a matroid was defined via the keyword `circuits` - `Matroid(circuits=[...])` - it was converted to a `BasisMatroid`. This required the computation of the bases upon definition which could be very slow.

<!-- Why is this change required? What problem does it solve? -->
<!-- If this PR resolves an open issue, please link to it here. For example "Fixes #12345". -->
<!-- If your change requires a documentation PR, please link it appropriately. -->
For some relevant discussion, have a look at #36820 and #36956. This PR is my proposed resolution of #36820 (Note that this provides a workaround and not a direct implementation of the issue's request. In fact, I deem that a direct implementation is undesirable due to memory considerations; see #36956.)

### :memo: Checklist

<!-- Put an `x` in all the boxes that apply. -->
<!-- If your change requires a documentation PR, please link it appropriately -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!-- Feel free to remove irrelevant items. -->

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation accordingly.

gmou3 committed:
Merge remote-tracking branch 'sagemath/develop' into circuits_matroid

gmou3 committed:
Add CircuitsMatroid class

gmou3 committed:
Merge remote-tracking branch 'sagemath/develop' into circuits_matroid

gmou3 committed:
Small docstrings edits

1-line outputs and iterable to SetSystem

tscrim commented on src/sage/matroids/circuits_matroid.pyx:
```diff
@@ -0,0 +1,613 @@
+
+cdef class CircuitsMatroid(Matroid):
+    r"""
+    INPUT:
```
```suggestion
cdef class CircuitsMatroid(Matroid):
    r"""
    A matroid defined by its set of circuits.

    INPUT:
```

tscrim commented on src/sage/matroids/circuits_matroid.pyx:
```diff
@@ -0,0 +1,613 @@
+    - If the input is a matroid ``M``, return a ``CircuitsMatroid`` instance
+      representing ``M``.
+    - Otherwise, return a ``CircuitsMatroid`` instance based on the
+      ``groundset`` and ``circuits``.
```
```suggestion
```
There is no point as the output is clearly this class. This isn't uniform and very, very in Sage (and likely should be removed in other places).

tscrim commented on src/sage/matroids/circuits_matroid.pyx:
```diff
@@ -0,0 +1,613 @@
+    def __init__(self, M=None, groundset=None, circuits=None, nsc_defined=False):
+        """
+        Initialization of the matroid. See class docstring for full
+        documentation.
```
Needs a doctest; typically a `TestSuite(M).run()`.

gmou3 replied:
That was at the top of the file. Moved it inside __init__.

tscrim commented on src/sage/matroids/circuits_matroid.pyx:
```diff
@@ -0,0 +1,613 @@
+
+        OUTPUT: a frozenset; a maximal independent subset of ``X``
+        """
+        I = set(F)
```
```suggestion
        cdef set I = set(F)
```

tscrim commented on src/sage/matroids/circuits_matroid.pyx:
```diff
@@ -0,0 +1,613 @@
+            for C in self._k_C[i]:
+                if i <= len(I) and i > 0:
+                    if C <= I:
+                        for e in C:
```
```suggestion
                        if e in C:
```
otherwise this is not doing anything.

gmou3 replied:
This is a fast way of picking an arbitrary element from C. (I know it looks weird.)

tscrim replied:
You should not rely on scoped variables. If you just want to get an element in `C`, then do
```python
e = next(iter(C))
```

tscrim commented on src/sage/matroids/circuits_matroid.pyx:
```diff
@@ -0,0 +1,613 @@
+        OUTPUT: boolean
+        """
+        I = set(F)
+        s = len(F)
```
```suggestion
        cdef set I = set(F)
        cdef int s = len(F)
```

tscrim commented on src/sage/matroids/constructor.py:
```diff
@@ -302,33 +303,33 @@ def Matroid(groundset=None, data=None, **kwds):
+            True
+
+
+
```
```suggestion
            True

```

tscrim commented on src/sage/matroids/unpickling.pyx:
```diff
@@ -28,13 +28,15 @@ AUTHORS:
-from sage.matroids.circuit_closures_matroid cimport CircuitClosuresMatroid
 from sage.matroids.basis_matroid cimport BasisMatroid
-from sage.matroids.linear_matroid cimport LinearMatroid, RegularMatroid, BinaryMatroid, TernaryMatroid, QuaternaryMatroid
+from sage.matroids.circuit_closures_matroid cimport CircuitClosuresMatroid
```
Let's not needlessly change the order of things, especially to something just as random.

tscrim commented on src/sage/matroids/circuits_matroid.pyx:
```diff
@@ -0,0 +1,613 @@
+
+    cpdef is_valid(self) noexcept:
+        r"""
+        Test if self obeys the matroid axioms.
```
```suggestion
        Test if ``self`` obeys the matroid axioms.
```

tscrim commented on src/sage/matroids/circuits_matroid.pyx:
```diff
@@ -0,0 +1,613 @@
+
+        [Oxl2011]_, p. 327.
+        """
+        return min([i for i in self._k_C], default=float('inf'))
```
```suggestion
        return min(self._k_C, default=float('inf'))
```

tscrim commented on src/sage/matroids/circuits_matroid.pyx:
```diff
@@ -0,0 +1,613 @@
+                    NSC.append(C)
+        return NSC
+
+    def no_broken_circuits_sets(self, order=None):
```
This is an iterator, but the other methods (at least by a spot check) return lists. This should be a list for consistency (perhaps good to change to an iterator at some point though).

gmou3 replied:
Thanks, this needed some essential changes.

tscrim commented on src/sage/matroids/circuits_matroid.pyx:
```diff
@@ -0,0 +1,613 @@
+        Otherwise an error is raised.
+        """
+        I = set(F)
+        for C in self.circuits():
```
```suggestion
        cdef set I = set(F)
        for C in self._C:
```
(or perhaps `self.circuits_iterator()`?)

gmou3 replied:
For a `CircuitsMatroid` there is no need to use the `circuits_iterator` internally as the circuits are already in memory.

tscrim reviewed (changes requested):
In addition to these comments, *all* `def` and `cpdef` methods need doctests.

gmou3 committed:
Impovements suggested by tscrim

gmou3 committed:
Edit amount of empty lines

Co-authored-by: Travis Scrimshaw <clfrngrown@aol.com>

gmou3 committed:
Merge remote-tracking branch 'origin/circuits_matroid' into circuits_matroid

gmou3 committed:
Update circuits_matroid.p*

Change 'order' to 'ordering'; use 'ordering' for consistency with matroid.p*. Add nonspanning_circuits_iterator() and improve docstrings.

tscrim commented on src/sage/matroids/database_matroids.py:
```diff
@@ -2116,7 +2116,8 @@ def Spike(r, t=True, C3=[]):
         sage: M = matroids.Spike(3, False); M
-        Free 3-spike\t: Matroid of rank 3 on 6 elements with 20 bases
+        Free 3-spike\t: M \ {'t'}, where M is Matroid of rank 3 on 7 elements
+        with 3 non-spanning circuits
```
```suggestion
         with 3 non-spanning circuits
```
Just as a subtle nod to the fact it is a continuing doctest (this sort of thing is common in Sage's doctests; IMO, it should be more universal).

gmou3 replied:
I went ahead and did many docstring changes. I changed some example blocks having in mind a 95th column limit, and I made the REFERENCE blocks into one-liners where applicable.

tscrim commented on src/sage/matroids/circuits_matroid.pyx:
```diff
@@ -0,0 +1,676 @@
+            for j in self._k_C:
+                if i <= j:
+                    for C1 in self._k_C[i]:
+                        if len(C1) == 0:
```
```suggestion
                        if not C1:
```
A microoptimization.

tscrim commented on src/sage/matroids/circuits_matroid.pyx:
```diff
@@ -0,0 +1,676 @@
+                                break
+                            for e in C1 & C2:
+                                flag = False
+                                S = (set(C1) | set(C2)) - {e}
```
```suggestion
                            UC12 = set(C1) | set(C2))
                            for e in C1 & C2:
                                flag = False
                                S = UC12- {e}
```
So you don't recreate the union every loop.

tscrim commented on src/sage/matroids/circuits_matroid.pyx:
```diff
@@ -0,0 +1,676 @@
+        """
+        for i in self._k_C:
+            for j in self._k_C:
+                if i <= j:
```
```suggestion
        sizes = sorted(self._k_C)
        for ii, i in enumerate(sizes):
            for jj,j in enumerate(sizes[:ii+1]):
```
Fewer checks and simplifies the logic IMO. Actually, you probably can check if the first size is zero and then skip it to further simplify the next lines. The `jj` will be used below.

gmou3 replied:
Seems more complicated to me. These checks are nothing compared to the algorithm's complexity.

tscrim replied:
If you want to do something by triangularity, then I think you should do it as such. I agree that the checks are likely not large, there are a lot of them (approx $\binom{n}{2}$ here plus more later on). Although I find this code easier to read and understand with less indentation. (I also got the triangular order backwards; I updated my suggestion above to fix this.) If you really disagree, then please change this to
```python
if i > j:
    continue
```

tscrim commented on src/sage/matroids/circuits_matroid.pyx:
```diff
@@ -0,0 +1,676 @@
+                            if C1 < C2:
+                                return False
+                            if C1 == C2:
+                                break
```
```suggestion
```
This should never happen because `C1` should be smaller than `C2`, right?

gmou3 replied:
The size of `C1` is smaller or equal to that of `C2`.

tscrim replied:
I got the triangularity wrong. Yes, you're right.

tscrim commented on src/sage/matroids/circuits_matroid.pyx:
```diff
@@ -0,0 +1,676 @@
+                            for e in C1 & C2:
+                                flag = False
+                                S = (set(C1) | set(C2)) - {e}
+                                for k in self._k_C:
```
```suggestion
                                for k in sizes[jj+1:]:
```
It doesn't remove the check below, but it does mean you potentially get to skip a lot of checks.

gmou3 replied:
This is incorrect; the circuit can be smaller in size than `C2` (`jj`). Consider a cycle graph with one edge replaced by two parallel edges. Let `C1` and `C2` be the two big cycles with all edges in common except one. Then the `2`-cycle is the `C3` that we are looking for, and that is smaller than the size of `C2`.

tscrim replied:
Right, I was off by 1. The size of `S` is at least the size of `C2` (which is at least the size of `C1`) since it must be at least one larger than `C2` but then subtracting one for removing `e`. Anyways, moot point since we don't have `jj` with your recent change (which I don't oppose).

tscrim commented on src/sage/matroids/circuits_matroid.pyx:
```diff
@@ -0,0 +1,676 @@
+        for i in self._k_C:
+            if i <= self.rank():
+                for C in self._k_C[i]:
+                    NSC.append(C)
```
```suggestion
                NSC.extend(self._k_C[i])
```

tscrim commented on src/sage/matroids/circuits_matroid.pyx:
```diff
@@ -0,0 +1,676 @@
+            Iterator over a system of subsets
+        """
+        cdef SetSystem NSC
+        NSC = SetSystem(list(self.groundset()))
```
```suggestion
        cdef list NSC = list(self.groundset())
```

tscrim commented on src/sage/matroids/circuits_matroid.pyx:
```diff
@@ -0,0 +1,676 @@
+            if i <= self.rank():
+                for C in self._k_C[i]:
+                    NSC.append(C)
+        return NSC
```
```suggestion
        return SetSystem(NSC)
```

tscrim commented on src/sage/matroids/circuits_matroid.pyx:
```diff
@@ -0,0 +1,676 @@
+            sage: sum(1 for C in M.circuits_iterator())
+            4
+        """
+        if k:
```
```suggestion
        if k is not None:
```
`k` could be `0`, which would then not fall into this path. Probably good to add a test case for this.

I would actually explicitly list the circuits in the doctest too.

gmou3 replied:
Done. Good catch!

tscrim commented on src/sage/matroids/circuits_matroid.pyx:
```diff
@@ -0,0 +1,676 @@
+        """
+        cdef SetSystem C
+        C = SetSystem(list(self.groundset()))
+        if k:
```
```suggestion
        if k is not None:
```
Same as for the iterator.

tscrim commented on src/sage/matroids/circuits_matroid.pyx:
```diff
@@ -0,0 +1,676 @@
+                    flag = False
+                    break
+            if flag:
+                yield frozenset(B)
```
```suggestion
        cdef SetSystem NSC = self.nonspanning_circuits()
        for B in combinations(self._groundset, self._matroid_rank):
            B = frozenset(B)
            if not any(C <= B for C in NSC)
                yield B
```
For normal `def` functions you can use generators IIRC. Anyways, it is best to not recreate things many times over.

tscrim commented on src/sage/matroids/circuits_matroid.pyx:
```diff
@@ -0,0 +1,676 @@
+                    break
+            if flag:
+                B.append(S)
+        return B
```
```suggestion
        cdef SetSystem B, NSC
        cdef bint flag
        B = SetSystem(list(self.groundset()))
        NSC = self.nonspanning_circuits()
        from itertools import combinations
        for S in combinations(self._groundset, self._matroid_rank):
            flag = True
            S = frozenset(S)
            for C in NSC:
                if C <= S:
                    flag = False
                    break
            if flag:
                B.append(S)
        return B
```

tscrim commented on src/sage/matroids/circuits_matroid.pyx:
```diff
@@ -0,0 +1,676 @@
+        """
+        return self._matroid_rank
+
+    cpdef _is_independent(self, F) noexcept:
```
This method needs a doctest.

tscrim commented on src/sage/matroids/circuits_matroid.pyx:
```diff
@@ -0,0 +1,676 @@
+                        return False
+        return True
+
+    cpdef _max_independent(self, F) noexcept:
```
This method needs a doctest.

tscrim commented on src/sage/matroids/circuits_matroid.pyx:
```diff
@@ -0,0 +1,676 @@
+
+        return frozenset(I)
+
+    cpdef _circuit(self, F) noexcept:
```
This method needs a doctest.

tscrim commented on src/sage/matroids/circuits_matroid.pyx:
```diff
@@ -0,0 +1,676 @@
+                return C
+        raise ValueError("no circuit in independent set")
+
+    cpdef _is_isomorphic(self, other, certificate=False) noexcept:
```
This method needs a doctest.

tscrim commented on src/sage/matroids/circuits_matroid.pyx:
```diff
@@ -0,0 +1,676 @@
+
+    # representation
+
+    def _repr_(self):
```
This method needs a doctest.

tscrim reviewed (changes requested):
Thank you. It is looking good.

Here are some more comments. Some are necessary (again, all `def` and `cpdef` methods need doctests within their docstring), but others are some optimizations I think are fairly easy.

gmou3 committed:
More suggestions by tscrim and docstring improvements

More examples and other mostly minor improvements
Docstring improvements: 1-line references and outputs, 95 column limit in examples

tscrim commented on src/sage/matroids/circuits_matroid.pyx:
```diff
@@ -0,0 +1,737 @@
+        - ``certificate`` -- boolean (optional)
+
+        OUTPUT: boolean, and, if certificate = True, a dictionary giving the
+        isomorphism or None
```
```suggestion
        - ``certificate`` -- boolean (default: ``False``)

        OUTPUT: boolean, and, if ``certificate=True``, a dictionary giving the
        isomorphism or ``None``
```

tscrim reviewed (changes requested):
Thank you for the changes.

I think the way you had the references before was better (and it is more in line with how we do other things in Sage generally).

A few more specific comments.

gmou3 committed:
Less checks and indents in circuits is_valid()

gmou3 committed:
Revert formatting of references

tscrim commented on src/sage/matroids/circuits_matroid.pyx:
```diff
@@ -0,0 +1,740 @@
+                                for C3 in self._k_C[k]:
+                                    if C3 <= S:
+                                        flag = True
+                                        break
```
```suggestion
                    I12 = C1 & C2
                    if not I12:  # C1 and C2 are disjoint; nothing to test
                        continue
                    if len(I12) == len(C1):
                        if len(C1) == len(C2):  # they are equal
                            break
                        # a circuit can't be a subset of another circuit
                        return False 
                    # check circuit elimination axiom
                    UC12 = C1 | C2
                    Slen = len(UC12) - 1  # will be the size of S below
                    for e in I12:
                        flag = False
                        S = UC12 - {e}
                        for k in self._k_C:
                            if k <= Slen:
                                for C3 in self._k_C[k]:
                                    if C3 <= S:
                                        flag = True
                                        break
                            if flag:
                                break
```
This should short-curcuit properly to avoid extra checks. Since we expect that nearly all checks will not have containment, we can avoid repeating some iterations over the circuits by simply building the intersection first. I also don't see why we needed to create `set` copies of `C!` and `C2`.

Separately, it might be beneficial to
```
cdef int i, j, k, lenS
cdef bint flag
```
so the comparisons are done at the true C level.

gmou3 replied:
Looks good. Do you know if it is indeed better to add the `cdef i, j, k`, i.e. for variables defined in for loops? If so there are many other places I should do that.

tscrim replied:
It's good for comparing things that are `int`'s or castable into them. However, that involves some extra checks. The advantage is you get (much) faster comparisons between `int` objects. I am skeptical it will have any (net)  non-negligible speed benefits, so let's just leave it be.

gmou3 committed:
Optimizations suggested by tscrim

tscrim reviewed (changes requested):
Thank you. The only thing left is the order change in `unpickling.pyx`. I think you wanted alphabetical, but it is not. I also am necessarily not in favor of these input sorting-type changes; yet, I won't reject it here (provided it is resorted by some reasonable way).

gmou3 committed:
Edit unpickling.pyx imports

tscrim reviewed (approved):
Then let is be so.

For the future, please avoid changing the import orders. Alphabetical is not always best and these unrelated changes can create unnecessary conflicts.
